source/s2n/s2n_tls_channel_handler.c

```diff
@@ -577,29 +585,36 @@ static void s_s2n_pkcs11_async_pkey_task(
     struct aws_byte_buf output_buf; /* initialized later */
     AWS_ZERO_STRUCT(output_buf);
 
-    if (status != AWS_TASK_STATUS_RUN_READY) {
+    /* if things started failing since this task was scheduled, just clean up and bail out */
+    if (status != AWS_TASK_STATUS_RUN_READY || s2n_handler->negotiation_failed) {
         goto clean_up;
     }
 
-    // TODO: ensure connection/channel is still ok
+    AWS_LOGF_TRACE(AWS_LS_IO_TLS, "id=%p: Running PKCS#11 async pkey task", (void *)handler);
 
-    // TODO: log probably impossible s2n errors
+    /* We check all s2n_async_pkey_op functions for success,
+     * but they shouldn't fail if they're called correctly.
+     * Even if the output is bad, the failure will happen later in s2n_negotiate() */
 
     uint32_t input_size = 0;
     if (s2n_async_pkey_op_get_input_size(op, &input_size)) {
-        aws_raise_error(AWS_ERROR_INVALID_STATE);
+        AWS_LOGF_ERROR(AWS_LS_IO_TLS, "id=%p: Failed querying s2n async pkey op size", (void *)handler);
+        aws_raise_error(AWS_IO_TLS_ERROR_NEGOTIATION_FAILURE);
         goto error;
     }
 
     input_data = aws_mem_acquire(handler->alloc, input_size);
     if (s2n_async_pkey_op_get_input(op, input_data, input_size)) {
-        aws_raise_error(AWS_ERROR_INVALID_STATE);
+        AWS_LOGF_ERROR(AWS_LS_IO_TLS, "id=%p: Failed querying s2n async pkey input", (void *)handler);
+        aws_raise_error(AWS_IO_TLS_ERROR_NEGOTIATION_FAILURE);
```

Comment:
this seems more like an io error or a syscall error. would that be more instructive? to the user?

Reply:
I'll go back to INVALID_STATE.
Yeah, I wasn't sure what to do here. But talking with some s2n engineers, these should never fail unless we're doing something really terribly wrong